File: crawler/geo.py
import re
from typing import Optional


CHINA_LOCATION_MARKERS = (
    "china", "中国", "prc", "greater china",
    # ISO-2 国家码。外企 ATS（SmartRecruiters / Greenhouse 等）的 location.country 直接给
    # 小写国别码，城市名却是**空格分词的拼音**（"He Fei Shi" / "Ning Bo Shi" / "Zhe Jiang Sheng"），
    # 拼音表按 \b 匹配一个都对不上 —— 大陆集团 29 个中国岗里 8 个（28%）就这么被判成 None。
    # 认这个码是**对方自己声明的国别**，不是我们猜的；全库 596 行含独立 "cn" 词的 active 岗
    # 逐行核过，现判定 100% 已经是 CN，加它零误伤（2026-09-05 live 实测）。
    "cn",
    # First/new first-tier and major industrial cities. Foreign ATS boards often
    # provide only a city/province name without an explicit "China" suffix.
    "beijing", "shanghai", "shenzhen", "guangzhou", "hangzhou", "chengdu",
    "nanjing", "suzhou", "wuhan", "xi'an", "xian", "foshan", "dongguan",
    "tianjin", "chongqing", "wuxi", "ningbo", "qingdao", "dalian", "xiamen",
    "hefei", "changsha", "zhengzhou", "jinan", "kunming", "shijiazhuang",
    "changchun", "harbin", "shenyang", "nanchang", "fuzhou", "nanning",
    "guiyang", "lanzhou", "taiyuan", "wenzhou", "zhuhai", "yantai", "xuzhou",
    "changzhou", "nantong", "weifang", "luoyang", "huizhou",
    # Provinces/autonomous regions in pinyin.
    "jiangsu", "zhejiang", "guangdong", "sichuan", "shandong", "henan",
    "hebei", "hunan", "hubei", "anhui", "fujian", "jiangxi", "liaoning",
    "shaanxi", "shanxi", "yunnan", "guizhou", "gansu", "hainan", "jilin",
    "heilongjiang", "qinghai", "ningxia", "xinjiang", "guangxi",
    "nei mongol", "inner mongolia",
    "北京", "上海", "深圳", "广州", "杭州", "成都", "南京", "苏州", "武汉", "西安", "佛山",
    "天津", "重庆", "无锡", "宁波", "青岛", "大连", "厦门", "合肥", "长沙", "郑州",
    "hong kong", "香港", "macau", "macao", "澳门",
)

_CJK_MARKERS = tuple(m for m in CHINA_LOCATION_MARKERS if any("一" <= ch <= "鿿" for ch in m))
_LATIN_MARKERS = tuple(m for m in CHINA_LOCATION_MARKERS if m not in _CJK_MARKERS)
_LATIN_MARKER_RE = re.compile(r"\b(?:" + "|".join(re.escape(m) for m in _LATIN_MARKERS) + r")\b")

REMOTE_MARKERS = ("remote", "anywhere", "distributed", "work from home", "wfh", "远程", "远端")

OVERSEAS_LOCATION_TOKENS = {
    "usa", "us", "canada", "uk", "britain", "ireland", "germany", "france", "netherlands",
    "spain", "italy", "poland", "portugal", "sweden", "switzerland", "austria", "belgium",
    "europe", "emea", "americas", "latam", "brazil", "mexico", "argentina", "colombia",
    "india", "japan", "korea", "singapore", "malaysia", "thailand", "vietnam", "indonesia",
    "philippines", "australia", "nz", "uae", "dubai", "israel", "egypt", "turkey", "africa",
}
OVERSEAS_LOCATION_PHRASES = (
    "united states", "united kingdom", "new zealand", "south korea", "saudi arabia",
    "sri lanka", "costa rica", "south africa",
)

_COUNTRY_TOKENS = {
    # ⚠️ TW 必须排在 CN 前面（dict 按插入序遍历，先命中先返回）。
    # 台湾按项目口径**不抓、不归入任一范围**：TW ∉ 任何 source regions 且 TW ∉ _GREATER_CHINA，
    # 于是 location_in_scope 一律返回 False。
    # 此前 TW 压根不在本表里 → "Taipei, Taipei shih, Taiwan, Province of China" 这种写法
    # 因为含 "china" 被判成 CN 放行（2026-07-28 Siemens 改成翻全分页后实测捞进 5 个台北岗才暴露）。
    # 原有 test_taiwan_is_not_in_any_active_scope 只覆盖 "Taiwan"/"Taipei, Taiwan"/"台北, 台湾"
    # 这类**不含 china 字样**的写法（code=None 自然落 False），所以一直是绿的、盖不住这个洞。
    "TW": ["taiwan", "台湾", "臺灣", "taipei", "台北", "臺北", "kaohsiung", "高雄", "hsinchu", "新竹"],
    "HK": ["hong kong", "香港", "hongkong"],
    "MO": ["macau", "macao", "澳门"],
    "CN": [m for m in CHINA_LOCATION_MARKERS if m not in {"hong kong", "香港", "macau", "macao", "澳门"}],
    "US": [
        "united states", "usa", "u.s.", "u.s.a", "america", "us",
        "new york", "纽约", "san francisco", "旧金山", "sf bay", "bay area",
        "seattle", "西雅图", "sunnyvale", "mountain view", "cupertino", "san jose",
        "santa clara", "palo alto", "austin", "boston", "chicago", "los angeles",
        "washington", "atlanta", "denver", "dallas", "houston", "san diego",
        "redmond", "menlo park", ", ca", ", ny", ", wa", ", tx", ", ma",
    ],
    "SG": ["singapore", "sg", "新加坡"],
}
_GREATER_CHINA = {"CN", "HK", "MO"}


def _norm(text: Optional[str]) -> str:
    return (text or "").strip().lower()


def _contains_token(text: str, token: str) -> bool:
    if any("一" <= ch <= "鿿" for ch in token) or token.startswith(","):
        return token in text
    parts = [re.escape(p) for p in re.split(r"[^a-z0-9]+", token.lower()) if p]
    if not parts:
        return False
    pattern = r"[^a-z0-9]+".join(parts)
    return bool(re.search(r"(?<![a-z0-9])" + pattern + r"(?![a-z0-9])", text))


def is_china_location(location: Optional[str]) -> bool:
    """Whether a location belongs to greater China, including Hong Kong/Macau."""
    if not location:
        return False
    text = location.lower()
    if any(marker in text for marker in _CJK_MARKERS):
        return True
    if _LATIN_MARKER_RE.search(text):
        return True
    norm = re.sub(r"[\s,\-/]+", " ", text)
    return bool(_LATIN_MARKER_RE.search(norm))


def is_remote_location(location: Optional[str]) -> bool:
    if not location:
        return False
    return any(marker in location.lower() for marker in REMOTE_MARKERS)


def _is_overseas_pinned(location: Optional[str]) -> bool:
    if not location:
        return False
    text = location.lower()
    if any(phrase in text for phrase in OVERSEAS_LOCATION_PHRASES):
        return True
    tokens = [t for t in re.split(r"[^a-z]+", text) if t]
    return any(t in OVERSEAS_LOCATION_TOKENS for t in tokens)


def keep_for_china_radar(location: Optional[str]) -> bool:
    """Existing China radar scope: greater China plus remote not pinned overseas."""
    if is_china_location(location):
        return True
    if is_remote_location(location) and not _is_overseas_pinned(location):
        return True
    return False


def derive_country_code(location: Optional[str]) -> Optional[str]:
    """Derive an ISO-2 country/region code from free-form location text."""
    text = _norm(location)
    if not text or text in ("unknown", "multiple locations"):
        return None
    for code, tokens in _COUNTRY_TOKENS.items():
        if any(_contains_token(text, token) for token in tokens):
            return code
    return None


def derive_job_scope(location: Optional[str]) -> str:
    """domestic for greater China and location-less jobs; bare remote counts as overseas.

    ⚠️ 「没写国家的远程岗」曾一律算 domestic，等于把外企的远程岗当成中国供给：
    2026-09-05 live 实测全库 9,873 个 active 岗地点是光秃秃的「远程」，逐个核过
    **一个中国岗都没有** —— 前 30 家全是 AbbVie / ServiceNow / NVIDIA / Pfizer 这类外企，
    连唯一一家本土公司（腾讯 7 个）的 jd_url 里也明写着 Warsaw / Thailand / Vietnam。
    这批岗占 domestic 总量 327,086 的 3.0%，属「指标诚实」红线里的注水。

    同时这修好了一个**从来没生效过的筛选项**：lib/job-scope.ts 的 Remote 档要求
    `job_scope='overseas' and country_code is null`，而旧规则永远产不出这个组合 ——
    live 实测该组合 0 行，即用户勾「海外 + 远程」必然空手而归。

    地点为空/unknown 仍算 domestic：本土 ATS（moka 等）大量岗位不填地点，
    把它们推去海外会让国内看板凭空少掉一批真岗。
    """
    code = derive_country_code(location)
    if code is None:
        return "overseas" if is_remote_location(location) else "domestic"
    return "domestic" if code in _GREATER_CHINA else "overseas"


def location_in_scope(location: Optional[str], regions) -> bool:
    """Whether location is inside source regions such as CN/US/SG/Remote."""
    regions = {str(r) for r in (regions or {"CN"})}
    code = derive_country_code(location)
    if code is not None:
        if code in regions:
            return True
        if "CN" in regions and code in _GREATER_CHINA:
            return True
        return False
    if is_remote_location(location):
        return "Remote" in regions or "CN" in regions
    return False

File: lib/geo.js
const CHINA_LOCATION_MARKERS = [
  "china", "中国", "prc", "greater china",
  // ISO-2 国家码；与 crawler/geo.py 的 CHINA_LOCATION_MARKERS 逐条一致（改一边必须改另一边）。
  // 外企 ATS 的 location.country 给小写国别码，城市却是空格分词拼音（"He Fei Shi"），
  // 拼音表按词边界一个都对不上，中国岗会被当成非中国岗丢掉。
  // 全库含独立 "cn" 词的 active 岗现判定 100% 已是 CN，加它零误伤。
  "cn",
  "beijing", "shanghai", "shenzhen", "guangzhou", "hangzhou", "chengdu",
  "nanjing", "suzhou", "wuhan", "xi'an", "xian", "foshan", "dongguan",
  "tianjin", "chongqing", "wuxi", "ningbo", "qingdao", "dalian", "xiamen",
  "hefei", "changsha", "zhengzhou", "jinan", "kunming", "shijiazhuang",
  "changchun", "harbin", "shenyang", "nanchang", "fuzhou", "nanning",
  "guiyang", "lanzhou", "taiyuan", "wenzhou", "zhuhai", "yantai", "xuzhou",
  "changzhou", "nantong", "weifang", "luoyang", "huizhou",
  "jiangsu", "zhejiang", "guangdong", "sichuan", "shandong", "henan",
  "hebei", "hunan", "hubei", "anhui", "fujian", "jiangxi", "liaoning",
  "shaanxi", "shanxi", "yunnan", "guizhou", "gansu", "hainan", "jilin",
  "heilongjiang", "qinghai", "ningxia", "xinjiang", "guangxi",
  "nei mongol", "inner mongolia",
  "北京", "上海", "深圳", "广州", "杭州", "成都", "南京", "苏州", "武汉", "西安", "佛山",
  "天津", "重庆", "无锡", "宁波", "青岛", "大连", "厦门", "合肥", "长沙", "郑州",
  "hong kong", "香港", "macau", "macao", "澳门",
];

const COUNTRY_TOKENS = {
  HK: ["hong kong", "香港", "hongkong"],
  MO: ["macau", "macao", "澳门"],
  CN: CHINA_LOCATION_MARKERS.filter((m) => !["hong kong", "香港", "macau", "macao", "澳门"].includes(m)),
  US: [
    "united states", "usa", "u.s.", "u.s.a", "america", "us",
    "new york", "纽约", "san francisco", "旧金山", "sf bay", "bay area",
    "seattle", "西雅图", "sunnyvale", "mountain view", "cupertino", "san jose",
    "santa clara", "palo alto", "austin", "boston", "chicago", "los angeles",
    "washington", "atlanta", "denver", "dallas", "houston", "san diego",
    "redmond", "menlo park", ", ca", ", ny", ", wa", ", tx", ", ma",
  ],
  SG: ["singapore", "sg", "新加坡"],
};

const GREATER_CHINA = new Set(["CN", "HK", "MO"]);
const REMOTE_MARKERS = ["remote", "anywhere", "distributed", "work from home", "wfh", "远程", "远端"];
const CJK_RE = /[\u4e00-\u9fff]/;

function norm(text) {
  return String(text || "").trim().toLowerCase();
}

function escapeRegExp(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function containsToken(text, token) {
  if (CJK_RE.test(token) || token.startsWith(",")) {
    return text.includes(token);
  }
  const parts = token.toLowerCase().split(/[^a-z0-9]+/).filter(Boolean).map(escapeRegExp);
  if (!parts.length) {
    return false;
  }
  return new RegExp(`(?<![a-z0-9])${parts.join("[^a-z0-9]+")}(?![a-z0-9])`).test(text);
}

function isRemoteLocation(location) {
  const text = norm(location);
  return REMOTE_MARKERS.some((marker) => text.includes(marker));
}

function deriveCountryCode(location) {
  const text = norm(location);
  if (!text || text === "unknown" || text === "multiple locations") {
    return null;
  }
  for (const [code, tokens] of Object.entries(COUNTRY_TOKENS)) {
    if (tokens.some((token) => containsToken(text, token))) {
      return code;
    }
  }
  return null;
}

// sources.regions 归一：supabase-js 把 text[] 解析成数组，但 PG 数组字面量字符串 "{CN,US}"
// 也可能一路传到这里 —— 与 crawler/normalizer.source_regions 同口径地剥 {}、切逗号、trim。
// ⚠️ 不能直接 Array.from(字符串)：那会切成单个字符，"CN" 变 ["C","N"] 后判不出 CN。
function normalizeRegions(regions) {
  if (!regions) return [];
  let list;
  if (Array.isArray(regions)) {
    list = regions;
  } else if (typeof regions === "string") {
    const text = regions.trim();
    const inner = text.startsWith("{") && text.endsWith("}") ? text.slice(1, -1) : text;
    list = inner.split(",");
  } else {
    list = Array.from(regions);
  }
  return list.map((r) => String(r).trim()).filter(Boolean);
}

/**
 * domestic = 大中华（CN/HK/MO），其余 overseas。与 crawler/geo.py 的 derive_job_scope 同口径，
 * 改一处必须两处同改（tests/geo-cross-lang.test.js 会拿同一批用例对拍两端）。
 *
 * 地点**抽不出国家**时（空 / "Multiple Locations" / 裸「远程」「Remote」），按**源自己的
 * regions** 兜底判定；没传 regions 则维持旧默认 domestic，故老调用方行为一字不变。
 *
 * ⚠️ 为什么必须看源：裸「远程」默认判 domestic 是**海外扩展之前**的合理默认（那时库里只有
 * CN 源，远程岗几乎必然在国内）。2026-07-02 放开 US/SG/Remote 之后这个默认就反了 ——
 * 2026-09-05 香港库实测：「location 抽不出国家」的在招岗里，源含 CN 的 114,075 个全是
 * domestic、源不含 CN 的 16,159 个全是 overseas（Wells Fargo 4,419 / Target 3,670 /
 * JLL 1,945 / AbbVie 1,512…），交叉项 0 个 —— 按源判是干净的。
 * （带国家写法的远程「Remote - US」早已由 f306271 修好，这里补的是**裸远程**那一半。）
 */
function deriveJobScope(location, regions) {
  const code = deriveCountryCode(location);
  if (code) {
    return GREATER_CHINA.has(code) ? "domestic" : "overseas";
  }
  const scoped = normalizeRegions(regions);
  if (scoped.length && !scoped.includes("CN")) {
    return "overseas";
  }
  return "domestic";
}

function locationInScope(location, regions) {
  const regionSet = new Set((regions && regions.length ? regions : ["CN"]).map(String));
  const code = deriveCountryCode(location);
  if (code) {
    if (regionSet.has(code)) return true;
    if (regionSet.has("CN") && GREATER_CHINA.has(code)) return true;
    return false;
  }
  if (isRemoteLocation(location)) {
    return regionSet.has("Remote") || regionSet.has("CN");
  }
  return false;
}

module.exports = {
  deriveCountryCode,
  deriveJobScope,
  locationInScope,
  normalizeRegions,
};
